src/enclave/http.h

```diff
@@ -0,0 +1,250 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "tlsendpoint.h"
+
+#include <http-parser/http_parser.h>
+
+namespace enclave
+{
+  namespace http
+  {
+    std::vector<uint8_t> post(const std::string& body)
+    {
+      auto req = fmt::format(
+        "POST / HTTP/1.1\r\n"
+        "Content-Type: application/json\r\n"
+        "Content-Length: {}\r\n\r\n{}",
+        body.size(),
+        body);
+      return std::vector<uint8_t>(req.begin(), req.end());
+    }
+
+    std::vector<uint8_t> post_header(const std::vector<uint8_t>& body)
+    {
+      auto req = fmt::format(
+        "POST / HTTP/1.1\r\n"
+        "Content-Type: application/json\r\n"
+        "Content-Length: {}\r\n\r\n{}",
+        body.size());
+      return std::vector<uint8_t>(req.begin(), req.end());
+    }
+
+    class MsgProcessor
+    {
+    public:
+      virtual void msg(std::vector<uint8_t> m) = 0;
+    };
+
+    enum State
+    {
+      DONE,
+      IN_MESSAGE
+    };
+
+    static int on_msg(http_parser* parser);
+    static int on_msg_end(http_parser* parser);
+    static int on_req(http_parser* parser, const char* at, size_t length);
+
+    class Parser
+    {
+    private:
+      http_parser parser;
+      http_parser_settings settings;
+      MsgProcessor& proc;
+      State state = DONE;
+      std::vector<uint8_t> buf;
+
+    public:
+      Parser(http_parser_type type, MsgProcessor& caller) : proc(caller)
```

Comment:
`caller` is a bit weird here as it could get confused with the identity of the client used elsewhere (i.e. `frontend.h`).

Reply:
Sure, how about "proc"?